packages/editor/src/utils/editor.ts

```diff
@@ -25,53 +27,104 @@ export const createBlockHub: (
+      if (end) {
+        const { model, rect } = end;
+        page.captureSync();
+        const distanceToTop = Math.abs(rect.top - point.y);
+        const distanceToBottom = Math.abs(rect.bottom - point.y);
+        const ids = page.addSiblingBlocks(
+          model,
+          blocks,
+          distanceToTop < distanceToBottom ? 'before' : 'after'
+        );
 
-      if (props[0].flavour === 'affine:database') {
-        const service = await getServiceOrRegister(props[0].flavour);
-        service.initDatabaseBlock(page, model, ids[0]);
+        // database init basic structure
+        if (props.flavour === 'affine:database') {
+          const service = await getServiceOrRegister(props.flavour);
+          service.initDatabaseBlock(page, model, ids[0]);
+        }
+
+        if (ids.length) {
+          asyncFocusRichText(page, ids[0]);
+        }
+      }
+
+      if (editor.mode === 'page') {
+        tryUpdateFrameSize(page, 1);
+        return;
       }
 
-      if (ids.length === 1) {
-        asyncFocusRichText(page, ids[0]);
+      // In edgeless mode.
+      const edgelessPageBlock = getEdgelessPage(page);
+      assertExists(edgelessPageBlock);
+
+      // Creates new frame block.
+      if (!end) {
+        const ids = edgelessPageBlock.addNewFrame(
+          blocks,
+          new Point(e.clientX, e.clientY)
+        );
+        if (ids.length) {
+          asyncFocusRichText(page, ids[0]);
+        }
       }
-      tryUpdateFrameSize(page, 1);
+
+      tryUpdateFrameSize(page, edgelessPageBlock.surface.viewport.zoom);
     },
-  });
+    onDragStarted: () => {
+      if (editor.mode === 'page') {
+        const defaultPageBlock = editor.querySelector('affine-default-page');
+        assertExists(defaultPageBlock);
+        defaultPageBlock.slots.selectedRectsUpdated.emit([]);
+      }
+    },
+    getAllowedBlocks: () => {
+      if (editor.mode === 'page') {
+        const defaultPageBlock = editor.querySelector('affine-default-page');
+        assertExists(defaultPageBlock);
+        return getAllowSelectedBlocks(defaultPageBlock.model);
+      } else {
+        const edgelessPageBlock = editor.querySelector('affine-edgeless-page');
+        assertExists(edgelessPageBlock);
+        return getAllowSelectedBlocks(edgelessPageBlock.model);
+      }
+    },
+    getHoveringFrameState: (point: Point) => {
+      const state = {
+        rect: null,
+        scale: 1,
+      } as { rect: Rect | null; scale: number };
 
-  if (editor.mode === 'page') {
-    const defaultPageBlock = editor.querySelector('affine-default-page');
-    assertExists(defaultPageBlock);
-    blockHub.slots = defaultPageBlock.slots;
-    blockHub.getAllowedBlocks = () =>
-      getAllowSelectedBlocks(defaultPageBlock.model);
-  } else {
-    const edgelessPageBlock = editor.querySelector('affine-edgeless-page');
-    assertExists(edgelessPageBlock);
-    blockHub.getAllowedBlocks = () =>
-      getAllowSelectedBlocks(edgelessPageBlock.model);
-  }
```

Comment:
This code will only be executed once. 

May need an update subscription when switching the mode.

/cc @doodlewind